matrix/plugins/glitch/main.py

```diff
@@ -0,0 +1,97 @@
+import asyncio
+import functools
+import logging
+import time
+import yaml
+
+from .selectors import Selectors
+from .actions import Actions
+from .plan import generate_plan, validate_plan
+
+
+log = logging.getLogger("glitch")
+DEFAULT_PLAN_NAME = "glitch_plan.{}.yaml"
+
+
+def default_resolver(model, kind, name):
+    if kind not in ["application", "unit", "model", "controller", "relation"]:
+        return None
+    entities = getattr(model, kind + "s")
+    obj = entities[name]
+    return obj
+
+
+# XXX: this most likely will need an async def
+# depending on libjuju
+def select(model, selectors, objects=None, resolver=default_resolver):
+    if not selectors:
+        if objects is None:
+            # TODO: custom Exception class
+            raise ValueError('No valid objects specified by selectors')
+        return objects
+
+    # if there are string names being passed (from a serialized plan for
+    # example) we must resolve them relative to the current model. This is
+    # pluggable using a resolver object which takes a model,
+    cur = None
+    args = [model]
+    # This can raise many an exception
+    for selector in selectors:
+        data = selector.copy()
+        m = Selectors.get(data.pop('selector'))
+        for k, v in data.items():
+            if isinstance(v, str):
+                # attempt resolution
+                o = resolver(model, k, v)
+                if o is not None:
+                    data[k] = o
+
+            #print(m, args, data)
+            cur = m(*args, **data)
+            args = [model, cur]
```

Comment:
It does do something: it adds the working set to the positional params.  Currently, "top-level" selectors like `units` don't actually accept an initial working set; thus, the `objects` list here is ignored.  I think making the working set explicit, required, and initialized by `objects` is a good idea.